### PR TITLE
sql/distsqlrun: avoid repeated EvalContext allocation

### PR DESCRIPTION
**Pull request by petermattis**
```
name        old time/op  new time/op  delta
Distinct-8   430µs ± 1%   291µs ± 2%  -32.31%  (p=0.000 n=9+10)
```

Release note (performance improvement): Speed up DISTINCT by avoiding an
unnecessary allocation.

### CODE REVIEW

File: pkg/sql/distsqlrun/distinct_test.go [hunk 1]
```diff
@@ -169,3 +169,47 @@ func TestDistinct(t *testing.T) {
...
+		DistinctColumns: []uint32{0},
+	}
+	post := &PostProcessSpec{}
+
+	columnTypeInt := sqlbase.ColumnType{SemanticType: sqlbase.ColumnType_INT}
+	types := make([]sqlbase.ColumnType, numCols)
+	for i := 0; i < numCols; i++ {
+		types[i] = columnTypeInt
+	}
+	makeInputTable := func(numRows, numCols int) sqlbase.EncDatumRows {
```
Comment (richardwu) [inline pkg/sql/distsqlrun/distinct_test.go]: nit: while you're here, can you pull this out from `BenchmarkHashJoiner` too into `distsqlrun.utils_test.go`?

File: pkg/sql/distsqlrun/distinct_test.go [hunk 2]
```diff
@@ -169,3 +169,47 @@ func TestDistinct(t *testing.T) {
...
+	flowCtx := &FlowCtx{
+		Settings: cluster.MakeTestingClusterSettings(),
+		EvalCtx:  evalCtx,
+	}
+	spec := &DistinctSpec{
+		DistinctColumns: []uint32{0},
+	}
+	post := &PostProcessSpec{}
+
+	columnTypeInt := sqlbase.ColumnType{SemanticType: sqlbase.ColumnType_INT}
```
Comment (richardwu) [inline pkg/sql/distsqlrun/distinct_test.go]: nit: you could also `intType` from `distsqlrun/utils_test.go` (same goes for `BenchmarkHashJoiner`.

**Review by richardwu — APPROVED** [review]

File: pkg/sql/distsqlrun/distinct_test.go (hunk 2 above)
Comment (petermattis) [reply pkg/sql/distsqlrun/distinct_test.go]: Done.

File: pkg/sql/distsqlrun/distinct_test.go (hunk 1 above)
Comment (petermattis) [reply pkg/sql/distsqlrun/distinct_test.go]: Done.

**Review by petermattis — COMMENTED** [review]
TFYR

**Pull request merged**